Neos.Flow/Classes/Core/Booting/Scripts.php

```diff
@@ -736,9 +741,73 @@ public static function buildPhpCommand(array $settings)
             $command .= ' -c ' . escapeshellarg($useIniFile);
         }
 
+        self::ensureWebSubrequestsUseCurrentlyRunningPhpVersion($command);
+
         return $command;
     }
 
+    /**
+     * Compares the realpath of the configured PHP binary (if any) with the one flow was called with in a CLI request.
+     * This avoids config errors where users forget to set Neos.Flow.core.phpBinaryPathAndFilename in CLI.
+     *
+     * @param string phpBinaryPathAndFilename
+     * @throws Exception
+     */
+    protected static function ensureCLISubrequestsUseCurrentlyRunningPhpBinary($phpBinaryPathAndFilename)
+    {
+        // Do nothing for non-CLI requests
+        if (PHP_SAPI !== 'cli') {
+            return;
+        }
+
+        // Resolve any symlinks that the configured php might be pointing to
+        $configuredPhpBinaryPathAndFilename = realpath($phpBinaryPathAndFilename);
+
+        // if the configured PHP binary is empty here, the file does not exist. We ignore that here because it is checked later in the script.
+        if (strlen($configuredPhpBinaryPathAndFilename) === 0) {
+            return;
+        }
+
+        if (strcmp(PHP_BINARY, $configuredPhpBinaryPathAndFilename) !== 0) {
+            throw new Exception(sprintf('You are running the Flow CLI with a PHP binary different from the one Flow is configured to use internally. ' .
+                'Flow has been run with "%s", while the PHP version Flow is configured to use for subrequests is "%s". Make sure to configure Flow to ' .
+                'use the same PHP binary by setting the "Neos.Flow.core.phpBinaryPathAndFilename" configuration option to "%s". Flush the ' .
+                'caches by removing the folder Data/Temporary before running ./flow again.',
+                PHP_BINARY, $configuredPhpBinaryPathAndFilename, PHP_BINARY), 1536303119);
+        }
+    }
+
+    /**
+     * Compares the actual version of the configured PHP binary (if any) with the one flow was called with in a non-CLI request.
+     * This avoids config errors where users forget to set Neos.Flow.core.phpBinaryPathAndFilename in connection with a web
+     * server.
+     *
+     * @param string $phpCommand the completely build php string that is used to execute subrequests
+     * @throws Exception
+     */
+    protected static function ensureWebSubrequestsUseCurrentlyRunningPhpVersion($phpCommand)
+    {
+        // Do nothing for CLI requests
+        if (PHP_SAPI === 'cli') {
+            return;
+        }
+
+        exec($phpCommand . ' -r "echo phpversion();"', $output, $result);
+
+        if ($result !== 0) {
+            return;
+        }
+        $runningPHPVersion = phpversion();
+        $configuredPHPVersion = $output[0];
+        if ($runningPHPVersion !== $configuredPHPVersion) {
```

Comment:
This could be a quite intrusive limitations for some systems/ hosting environments.
Since PHP for web requests and PHP on cli are completely different binaries, in theory versions can also differ. E.g. PHP 7.1.3 for web requests and PHP 7.1.2 on CLI

With this change, such constellation would throw an exception. Right?

Reply:
I agree, I don't think we need to be that strict. I would rather loosen it to same minor or something. Ideally we would go into the direction of https://github.com/kitsunet/flownative-webcompiler at some point to avoid the CLI thing altogether.

Reply:
> I would rather loosen it to same minor or something

Yes, that's what we did in the Setup package. It's not a beauty, but it seems to work:
https://github.com/neos/setup/blob/master/Classes/Core/RequestHandler.php#L173-L271

> Ideally we would go into the direction of https://github.com/kitsunet/flownative-webcompiler

It would be great if compilation worked without additional requests (I hope this works memory wise, too). But CLI "sub requests" are useful anyways (i.e. for "poor mans" job queue, batched CLI commands, ...). So +1 for making this more stable anyways

Reply:
Yes, definitely needs to be less strict. Minor version match is enough.